### PR TITLE
Homework 5 task 2 save original info

### PR DESCRIPTION
**Pull request by ivanzuev78**
I have some problems.

### CODE REVIEW

File: HomeWork_5/test_task_2_save_original_info.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+
+@pytest.fixture
+def decorated_func():
+    @print_result
+    def original_func(*args):
+        """This function can sum any objects which have __add___"""
+        return functools.reduce(lambda x, y: x + y, args)
+
+    init_decorated_func = original_func(1, 2)
```
Comment (ivanzuev78) [inline HomeWork_5/test_task_2_save_original_info.py:14]: I have to call original_func to init it and rewrite __doc__ and __name__.
Is there any way to rewrite __doc__ and __name__ without calling original_func?

Comment (Cjkjvfnby) [reply HomeWork_5/test_task_2_save_original_info.py:14]: 1) use insert code action for function names  `__doc__` (click edit on this message to see markdown)
2) google your question

**Review by Cjkjvfnby — CHANGES_REQUESTED** [review]

File: HomeWork_5/test_task_2_save_original_info.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        == "This function can sum any objects which have __add___"
+    )
+
+
+def test_print_result_original_name_saved(decorated_func):
+    assert decorated_func.__name__ == "original_func"
+
+
+def test_print_result_original_func_saved(decorated_func, capsys):
+    result_from_original_func = decorated_func.__original_func(1, 2, 3, 4)
```
Comment (Cjkjvfnby) [inline HomeWork_5/test_task_2_save_original_info.py:36]: Is it the same function or a different one?

File: HomeWork_5/test_task_2_save_original_info.py (hunk 1 above)
Comment (ivanzuev78) [reply HomeWork_5/test_task_2_save_original_info.py:14]: 1. Thanks
2. Ok

File: HomeWork_5/test_task_2_save_original_info.py (hunk 2 above)
Comment (ivanzuev78) [reply HomeWork_5/test_task_2_save_original_info.py:36]: What function do you mean?
If called function, it is non-decorated `original_func`.

Comment (Cjkjvfnby) [reply HomeWork_5/test_task_2_save_original_info.py:35]: You have 
1. origin al function
1. decorated function
1. function that can be accessed by __original_func

Are 1 and 3 the same?  

For example, different functions can produce the same result.  
max(0, 0) == min(0, 0)

I task you need to check that 3 is 1.   And checking the result is not a good way to do it.

File: HomeWork_5/save_original_info.py [hunk 3]
```diff
@@ -1,26 +1,39 @@
...
-исходную функцию в атрибуте __original_func
+import functools
 
-print_result изменять нельзя, за исключением добавления вашего
-декоратора на строку отведенную под него - замените комментарий
 
-До применения вашего декоратор будет вызываться AttributeError при custom_sum.__original_func
-Это корректное поведение
-После применения там должна быть исходная функция
+def my_decorator(func):
```
Comment (Cjkjvfnby) [inline HomeWork_5/save_original_info.py:4]: Is it a good name?  Does it describe the purpose of this function?

**Review by Cjkjvfnby — CHANGES_REQUESTED** [review]

File: HomeWork_5/test_task_2_save_original_info.py (hunk 2 above)
Comment (ivanzuev78) [reply HomeWork_5/test_task_2_save_original_info.py:35]: Implementation was not obvious to me, but it is done.

File: HomeWork_5/save_original_info.py (hunk 3 above)
Comment (ivanzuev78) [reply HomeWork_5/save_original_info.py:4]: Of course not. Fixed.

File: HomeWork_5/test_task_2_save_original_info.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+def decorated_and_origin_funcs():
+    def original_func(*args):
+        """This function can sum any objects which have __add___"""
+        return functools.reduce(lambda x, y: x + y, args)
+
+    decorated_func = print_result(original_func)
+    yield decorated_func, original_func
+
+
+def test_print_result_original_func_saved(decorated_and_origin_funcs, capsys):
```
Comment (Cjkjvfnby) [inline HomeWork_5/test_task_2_save_original_info.py:44]: If you use fixture once you can just do it in the test.  And you can reuse some existiong function

```
def test_print_result_original_func_saved()
    decorated = print_result(sum)
    assert decorated_func.__original_func is sum
```

**Review by Cjkjvfnby — APPROVED** [review]

File: HomeWork_5/test_task_2_save_original_info.py (hunk 4 above)
Comment (ivanzuev78) [reply HomeWork_5/test_task_2_save_original_info.py:44]: accepted.

**Pull request merged**